prov/verbs/src/windows/verbs_nd_init.c

```diff
@@ -0,0 +1,751 @@
+	nd_infra.adapters.count =
+		unique(nd_infra.adapters.adapter, nd_infra.adapters.count,
+		       sizeof(struct nd_adapter), nd_adapter_cmp);
+
+	// Allocate memory for adapter_list, device_list, and context_list all at once.
+	nd_infra.adapters.adapter_list =
+		(struct nd_adapter **)calloc((nd_infra.adapters.count + 1) * 3,
+					     sizeof(nd_infra.adapters.adapter));
+	if (!nd_infra.adapters.adapter_list)
+		return ND_NO_MEMORY;
+
+	nd_infra.adapters.device_list =
+		(struct ibv_device **)&nd_infra.adapters
+			.adapter_list[nd_infra.adapters.count + 1];
+	nd_infra.adapters.context_list =
+		(struct ibv_context **)&nd_infra.adapters
+			.adapter_list[(nd_infra.adapters.count + 1) * 2];
+
+	for (size_t i = 0; i < nd_infra.adapters.count; i++) {
+		adapter = &nd_infra.adapters.adapter[i];
+		nd_infra.adapters.adapter_list[i] = adapter;
+		nd_infra.adapters.device_list[i] = &adapter->device;
+		nd_infra.adapters.context_list[i] = &adapter->context;
+		factory = adapter->factory;
+
+		assert(factory);
+		assert(factory->provider);
+
+		assert(adapter->address.addr.sa_family == AF_INET ||
+		       adapter->address.addr.sa_family == AF_INET6);
+
+		hr = factory->provider->lpVtbl->ResolveAddress(
+			factory->provider, &adapter->address.addr,
+			(ULONG)ofi_sizeofaddr(&adapter->address.addr), &id);
+
+		if (FAILED(hr))
+			return hr;
+
+		hr = factory->provider->lpVtbl->OpenAdapter(
+			factory->provider, &IID_IND2Adapter, id,
+			(void **)&adapter->adapter);
+		if (FAILED(hr))
+			return hr;
+
+		adapter->info.InfoVersion = ND_VERSION_2;
+		hr = adapter->adapter->lpVtbl->Query(adapter->adapter,
+						     &adapter->info, &linfo);
+		if (FAILED(hr) && hr == ND_BUFFER_OVERFLOW) {
+			info = (ND2_ADAPTER_INFO *)malloc(linfo);
+			if (!info)
+				return ND_NO_MEMORY;
+			info->InfoVersion = ND_VERSION_2;
+			hr = adapter->adapter->lpVtbl->Query(adapter->adapter,
+							     info, &linfo);
+			if (FAILED(hr))
+				return hr;
+			adapter->info = *info;
+			free(info);
+		} else if (FAILED(hr)) {
+			return hr;
+		}
+
+		/* generate adapter's name */
+		addrlen = 0;
+		res = WSAAddressToStringW(
+			&adapter->address.addr,
+			(ULONG)ofi_sizeofaddr(&adapter->address.addr), NULL,
+			NULL, &addrlen);
+		if (res == SOCKET_ERROR && WSAGetLastError() == WSAEFAULT &&
+		    addrlen) {
+			saddr = (wchar_t *)malloc((addrlen + 1) *
+						  sizeof(*saddr));
+			WSAAddressToStringW(
+				&adapter->address.addr,
+				(ULONG)ofi_sizeofaddr(&adapter->address.addr),
+				NULL, saddr, &addrlen);
+		} else {
+			saddr = _wcsdup(L"unknown");
+		}
+
+		asprintf((char **)&adapter->name, "%S", saddr);
+		free(saddr);
+
+		adapter->device.transport_type = IBV_TRANSPORT_IB;
+		strncpy(&adapter->device.name[0], adapter->name,
+			min(strlen(adapter->name) + 1,
+			    sizeof(adapter->device.name)));
+		adapter->device.name[IBV_SYSFS_NAME_MAX - 1] = 0;
+
+		adapter->context.device = &adapter->device;
+		adapter->context.num_comp_vectors = 0;
+	}
+
+	return S_OK;
```

Comment:
This function has a lot of memory allocations and early exit paths.  Please use a common exit path for errors, so that cleanup of all allocations is handled properly.